File: chibi-api.mjs
/**
 * Chibi Pipeline API Bridge  —  localhost:5555
 * Node.js replacement for chibi-api.py (no Python required)
 *
 * Usage:   node chibi-api.mjs
 *
 * Endpoints
 *   POST /comfyui/upload      — upload image to ComfyUI input folder
 *   POST /comfyui/queue       — queue a workflow prompt
 *   GET  /comfyui/status/:id  — check job status / history
 *   GET  /comfyui/view        — proxy output image (?filename=X&type=Y&subfolder=Z)
 *   POST /rembg               — remove background (uses @imgly/background-removal-node)
 *   GET  /ping                — health check
 */

import http  from 'http';
import { URL } from 'url';
import { Buffer } from 'buffer';

const COMFYUI = 'http://localhost:8188';
const PORT    = 5555;

const CORS = {
  'Access-Control-Allow-Origin':  '*',
  'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type',
};

// ── helpers ────────────────────────────────────────────────────────────────

function reply(res, status, body, type = 'application/json') {
  const data = typeof body === 'string' ? Buffer.from(body) : Buffer.from(JSON.stringify(body));
  res.writeHead(status, { ...CORS, 'Content-Type': type, 'Content-Length': data.length });
  res.end(data);
}

function err(res, msg) {
  console.error('  [ERR]', msg);
  reply(res, 500, { error: String(msg) });
}

function readBody(req) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    req.on('data', c => chunks.push(c));
    req.on('end',  () => resolve(Buffer.concat(chunks)));
    req.on('error', reject);
  });
}

function httpGet(url, timeoutMs = 15000) {
  return new Promise((resolve, reject) => {
    const req = http.get(url, res => {
      const chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end',  () => resolve({ data: Buffer.concat(chunks), ct: res.headers['content-type'] || 'application/octet-stream' }));
    });
    req.setTimeout(timeoutMs, () => { req.destroy(); reject(new Error('Timeout')); });
    req.on('error', reject);
  });
}

function httpPost(urlStr, body, contentType, timeoutMs = 30000) {
  return new Promise((resolve, reject) => {
    const u = new URL(urlStr);
    const options = {
      hostname: u.hostname, port: u.port || 80,
      path: u.pathname + u.search, method: 'POST',
      headers: { 'Content-Type': contentType, 'Content-Length': body.length },
    };
    const req = http.request(options, res => {
      const chunks = [];
      res.on('data', c => chunks.push(c));
      res.on('end',  () => resolve({ data: Buffer.concat(chunks), ct: res.headers['content-type'] || 'application/json' }));
    });
    req.setTimeout(timeoutMs, () => { req.destroy(); reject(new Error('Timeout')); });
    req.on('error', reject);
    req.write(body);
    req.end();
  });
}

// ── route handlers ─────────────────────────────────────────────────────────

async function handleUpload(req, res) {
  try {
    const body     = await readBody(req);
    const payload  = JSON.parse(body.toString());
    const raw      = payload.imageData.includes(',') ? payload.imageData.split(',')[1] : payload.imageData;
    const imgBuf   = Buffer.from(raw, 'base64');
    const filename = payload.filename || `chibi_${Date.now()}.png`;

    const boundary = `----NodeBridge${Date.now()}`;
    const head     = Buffer.from(
      `--${boundary}\r\nContent-Disposition: form-data; name="image"; filename="${filename}"\r\nContent-Type: image/png\r\n\r\n`
    );
    const tail = Buffer.from(`\r\n--${boundary}--\r\n`);
    const form = Buffer.concat([head, imgBuf, tail]);

    const result = await httpPost(
      `${COMFYUI}/upload/image`, form,
      `multipart/form-data; boundary=${boundary}`
    );
    reply(res, 200, result.data.toString(), 'application/json');
  } catch (e) { err(res, e); }
}

async function handleQueue(req, res) {
  try {
    const body   = await readBody(req);
    const result = await httpPost(`${COMFYUI}/prompt`, body, 'application/json');
    reply(res, 200, result.data.toString(), 'application/json');
  } catch (e) { err(res, e); }
}

async function handleStatus(req, res, promptId) {
  try {
    const { data } = await httpGet(`${COMFYUI}/history/${promptId}`);
    reply(res, 200, data.toString(), 'application/json');
  } catch (e) { err(res, e); }
}

async function handleView(req, res, query) {
  try {
    const filename  = query.get('filename')  || '';
    const type      = query.get('type')      || 'output';
    const subfolder = query.get('subfolder') || '';
    const url = `${COMFYUI}/view?filename=${encodeURIComponent(filename)}&type=${type}&subfolder=${encodeURIComponent(subfolder)}`;
    const { data, ct } = await httpGet(url);
    reply(res, 200, data, ct);
  } catch (e) { err(res, e); }
}

async function handleRembg(req, res) {
  let removeBackground;
  try {
    ({ removeBackground } = await import('@imgly/background-removal-node'));
  } catch {
    err(res, 'Background removal package not found. Run: npm install @imgly/background-removal-node');
    return;
  }

  try {
    const body    = await readBody(req);
    const payload = JSON.parse(body.toString());
    const raw     = payload.imageData.includes(',') ? payload.imageData.split(',')[1] : payload.imageData;
    const imgBuf  = Buffer.from(raw, 'base64');

    console.log('  [rembg] Removing background… (first run downloads models ~50MB)');
    const resultBlob = await removeBackground(imgBuf);
    const resultBuf  = Buffer.from(await resultBlob.arrayBuffer());
    const b64        = resultBuf.toString('base64');
    reply(res, 200, { imageData: `data:image/png;base64,${b64}` });
    console.log('  [rembg] Done.');
  } catch (e) { err(res, e); }
}

// ── server ─────────────────────────────────────────────────────────────────

const server = http.createServer(async (req, res) => {
  const u     = new URL(req.url, `http://localhost:${PORT}`);
  const path  = u.pathname;
  const query = u.searchParams;

  if (req.method === 'OPTIONS') { reply(res, 200, ''); return; }

  try {
    if (req.method === 'GET') {
      if (path === '/ping') {
        reply(res, 200, { status: 'ok', comfyui: COMFYUI, node: process.version });
      } else if (path.startsWith('/comfyui/status/')) {
        const id = path.split('/').pop();
        await handleStatus(req, res, id);
      } else if (path === '/comfyui/view') {
        await handleView(req, res, query);
      } else {
        reply(res, 404, { error: 'Not found' });
      }
    } else if (req.method === 'POST') {
      if      (path === '/comfyui/upload') await handleUpload(req, res);
      else if (path === '/comfyui/queue')  await handleQueue(req, res);
      else if (path === '/rembg')          await handleRembg(req, res);
      else reply(res, 404, { error: 'Not found' });
    } else {
      reply(res, 405, { error: 'Method not allowed' });
    }
  } catch (e) {
    err(res, e);
  }
});

server.listen(PORT, 'localhost', () => {
  console.log('\n  ✦  Chibi Pipeline API Bridge (Node.js)');
  console.log(`     Running on  http://localhost:${PORT}`);
  console.log(`     ComfyUI at  ${COMFYUI}`);
  console.log(`     Node.js     ${process.version}\n`);
  console.log('  Endpoints:');
  console.log('     POST /comfyui/upload      — upload image to ComfyUI');
  console.log('     POST /comfyui/queue       — queue a workflow');
  console.log('     GET  /comfyui/status/:id  — check job status');
  console.log('     GET  /comfyui/view        — retrieve output image');
  console.log('     POST /rembg               — remove background');
  console.log('\n  Press Ctrl+C to stop.\n');
});

server.on('error', e => {
  if (e.code === 'EADDRINUSE') {
    console.error(`  Port ${PORT} is already in use. Stop the other process first.`);
  } else {
    console.error('  Server error:', e);
  }
  process.exit(1);
});

File: chibi-api.py
#!/usr/bin/env python3
"""
Chibi Pipeline API Bridge  —  localhost:5555
Bridges the Chibi Studio HTML tool to:
  - ComfyUI REST API (localhost:8188)
  - rembg background removal

Usage:
  python chibi-api.py

Dependencies:
  - rembg (for background removal): pip install rembg[gpu]
  - Pillow:                          pip install Pillow
  - Everything else is Python stdlib
"""

import json
import urllib.request
import urllib.parse
import base64
import io
import sys
import os
import uuid
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs

COMFYUI_URL = "http://localhost:8188"
PORT = 5555

# ── CORS headers ─────────────────────────────────────────────────────────────

CORS_HEADERS = {
    "Access-Control-Allow-Origin":  "*",
    "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
    "Access-Control-Allow-Headers": "Content-Type",
}


class Handler(BaseHTTPRequestHandler):

    # ── Routing ───────────────────────────────────────────────────────────────

    def do_OPTIONS(self):
        self._send(200, b"", "text/plain")

    def do_GET(self):
        parsed = urlparse(self.path)
        path   = parsed.path
        params = parse_qs(parsed.query)

        if path.startswith("/comfyui/status"):
            prompt_id = path.rstrip("/").split("/")[-1]
            self._proxy_get(f"{COMFYUI_URL}/history/{prompt_id}")

        elif path == "/comfyui/view":
            filename  = params.get("filename",  [""])[0]
            ftype     = params.get("type",      ["output"])[0]
            subfolder = params.get("subfolder", [""])[0]
            url = (f"{COMFYUI_URL}/view"
                   f"?filename={urllib.parse.quote(filename)}"
                   f"&type={ftype}"
                   f"&subfolder={urllib.parse.quote(subfolder)}")
            self._proxy_image(url)

        elif path == "/ping":
            self._json({"status": "ok", "comfyui": COMFYUI_URL})

        else:
            self._send(404, b"Not found", "text/plain")

    def do_POST(self):
        length = int(self.headers.get("Content-Length", 0))
        body   = self.rfile.read(length)
        path   = urlparse(self.path).path

        if   path == "/comfyui/upload": self._comfyui_upload(body)
        elif path == "/comfyui/queue":  self._comfyui_queue(body)
        elif path == "/rembg":          self._remove_bg(body)
        else: self._send(404, b"Not found", "text/plain")

    # ── ComfyUI: upload image ─────────────────────────────────────────────────

    def _comfyui_upload(self, body):
        try:
            data      = json.loads(body)
            raw_b64   = data["imageData"].split(",")[-1]
            img_bytes = base64.b64decode(raw_b64)
            filename  = data.get("filename", f"chibi_input_{uuid.uuid4().hex[:8]}.png")

            boundary  = b"----LikhaEraBoundary"
            form  = b"--" + boundary + b"\r\n"
            form += (f'Content-Disposition: form-data; name="image"; filename="{filename}"\r\n').encode()
            form += b"Content-Type: image/png\r\n\r\n"
            form += img_bytes
            form += b"\r\n--" + boundary + b"--\r\n"

            req = urllib.request.Request(
                f"{COMFYUI_URL}/upload/image",
                data=form,
                headers={"Content-Type": f"multipart/form-data; boundary={boundary.decode()}"},
            )
            with urllib.request.urlopen(req, timeout=30) as resp:
                result = resp.read()
            self._send(200, result, "application/json")

        except Exception as e:
            self._error(f"Upload failed: {e}")

    # ── ComfyUI: queue workflow ───────────────────────────────────────────────

    def _comfyui_queue(self, body):
        try:
            req = urllib.request.Request(
                f"{COMFYUI_URL}/prompt",
                data=body,
                headers={"Content-Type": "application/json"},
            )
            with urllib.request.urlopen(req, timeout=30) as resp:
                result = resp.read()
            self._send(200, result, "application/json")
        except Exception as e:
            self._error(f"Queue failed: {e}")

    # ── ComfyUI: proxy GET (history) ──────────────────────────────────────────

    def _proxy_get(self, url):
        try:
            with urllib.request.urlopen(url, timeout=15) as resp:
                data = resp.read()
            self._send(200, data, "application/json")
        except Exception as e:
            self._error(f"Proxy GET failed: {e}")

    # ── ComfyUI: proxy image ──────────────────────────────────────────────────

    def _proxy_image(self, url):
        try:
            with urllib.request.urlopen(url, timeout=15) as resp:
                data         = resp.read()
                content_type = resp.headers.get("Content-Type", "image/png")
            self._send(200, data, content_type)
        except Exception as e:
            self._error(f"Image proxy failed: {e}")

    # ── rembg: background removal ─────────────────────────────────────────────

    def _remove_bg(self, body):
        try:
            from rembg import remove
            from PIL import Image
        except ImportError:
            self._error(
                "rembg not installed.  Run:  pip install rembg[gpu]\n"
                "(CPU-only:  pip install rembg)"
            )
            return

        try:
            data      = json.loads(body)
            raw_b64   = data["imageData"].split(",")[-1]
            img_bytes = base64.b64decode(raw_b64)

            input_img  = Image.open(io.BytesIO(img_bytes)).convert("RGBA")
            output_img = remove(input_img)

            buf = io.BytesIO()
            output_img.save(buf, format="PNG")
            result_b64 = base64.b64encode(buf.getvalue()).decode()

            self._json({"imageData": f"data:image/png;base64,{result_b64}"})

        except Exception as e:
            self._error(f"Background removal failed: {e}")

    # ── helpers ───────────────────────────────────────────────────────────────

    def _send(self, code, body, content_type):
        self.send_response(code)
        for k, v in CORS_HEADERS.items():
            self.send_header(k, v)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def _json(self, obj):
        self._send(200, json.dumps(obj).encode(), "application/json")

    def _error(self, msg):
        print(f"  [ERROR] {msg}", file=sys.stderr)
        self._send(500, json.dumps({"error": msg}).encode(), "application/json")

    def log_message(self, fmt, *args):
        print(f"  [chibi-api] {fmt % args}")


# ── entry point ───────────────────────────────────────────────────────────────

if __name__ == "__main__":
    server = HTTPServer(("localhost", PORT), Handler)
    print(f"\n  ✦  Chibi Pipeline API Bridge")
    print(f"     Running on  http://localhost:{PORT}")
    print(f"     ComfyUI at  {COMFYUI_URL}")
    print(f"\n  Endpoints:")
    print(f"     POST /comfyui/upload   — upload image to ComfyUI input folder")
    print(f"     POST /comfyui/queue    — queue a workflow prompt")
    print(f"     GET  /comfyui/status/<id>  — check job status")
    print(f"     GET  /comfyui/view?filename=X&type=Y  — retrieve output image")
    print(f"     POST /rembg            — remove background (needs: pip install rembg)")
    print(f"\n  Press Ctrl+C to stop.\n")
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n  Stopped.")
